### ISSUE TITLE
"remote exception: grain supervisor is dead"

### ISSUE THREAD

**Issue by elimisteve** [sandstorm-io/sandstorm#2218]
```
HTTP 500 from https://api-168ddaae....oasis.sandstorm.io/rows; response: `Error: Error: remote exception: grain supervisor is dead
C++ location:(javascript):??
type: disconnected
```

Any idea what causes this?

Context: I'm hammering the CrypTag Sandstorm app pretty hard from my client app trying to recreate the nginx 504 Gateway issue that I experienced the other day that @kentonv said was probably related to the June 12th update and the fact that the CrypTag Sandstorm app was built at about the same time and probably is messed up. I upgraded the CrypTag Sandstorm web app and packaged it with the latest Sandstorm version, but I still got the error. So I created a custom version that turns DEBUG on, but I haven't run into the 504 again... but I _did_ run into the except above. Thanks.

**Comment by kentonv** [comment]
@elimisteve So, this is embarrassing, but we have a bug this week where accepting an app update notification does not actually update the app. So it's possible that you saw the problem continue to occur because the app hadn't actually been updated -- unless you manually uploaded your update or manually went back to the app market and clicked "install".

We've pushing an update this weekend which will correct the problem and re-notify people of missed updates.

**Comment by kentonv** [comment]
As for the new error, my first guess is that Sandstorm isn't treating HTTP API requests as keep-alives for the grain, so the grain shuts itself down after a couple minutes even if it is receiving API requests. The grain will immediately start back up on the next request, but if you catch it at just the right time, maybe it produces this error? This is just a guess, but if correct then this is in fact a Sandstorm bug.

**Comment by elimisteve** [comment]
@kentonv Re: your first comment -- okay that makes perfect sense, because I manually uploaded a new `.spk` file to my local Sandstorm Vagrant VM for testing, and I couldn't replicate the error that I was still getting from Oasis, despite having v0.8.2 of CrypTag accepted into the Sandstorm App Store.

Though I guess what I'm saying is only true if the bug you mention didn't apply to manually uploaded apps/SPKs. Is that the case?

**Comment by zarvox** [comment]
Yes, the bug only affected automatic updates being applied via the button in the notification.  Manually-uploaded SPKs or even manually installing from the App Market would have successfully installed the new package.

**Comment by kentonv** [comment]
@elimisteve Right, manually-uploaded SPKs -- or even going to the app store and clicking "install" -- worked correctly. It's only the auto-update notification that didn't work.

Really sorry about this. In retrospect I should have notified you about the updater bug.

**Comment by paulproteus** [comment]
If the theory from @kentonv is plausible, should we close this issue? @elimisteve curious for your input.

**Comment by elimisteve** [comment]
@paulproteus I just re-upped the frequency at which I grab data from Sandstorm and can confirm that the below message just reappeared, which I think confirms @kentonv's theory. I can now also confirm that the grain doesn't actually die or anything; I am successfully grabbing data from it even after getting this message (on the next fetch).

So for my use case, this error is harmless, because I'm polling quickly (and of course I could retry and it'd work); I thought it was related to a separate issue I was having (the Nginx 504 Gateway error), but it is not.

```
HTTP 500 from https://api-bdf205e4....oasis.sandstorm.io/...; response: `Error: remote exception: grain supervisor is dead
C++ location:(javascript):??
type: disconnected`
```

**Comment by paulproteus** [comment]
Got it, @elimisteve - so it meets this description:

> As for the new error, my first guess is that Sandstorm isn't treating HTTP API requests as keep-alives for the grain, so the grain shuts itself down after a couple minutes even if it is receiving API requests. The grain will immediately start back up on the next request, but if you catch it at just the right time, maybe it produces this error? This is just a guess, but if correct then this is in fact a Sandstorm bug.

Given that, I'll leave it open!

**Comment by elimisteve** [comment]
@paulproteus Exactly; that sounds like what's happening in my case. FYI, I don't see the error when I do 1 query/second, but for certain tests I've done 10 queries/sec, and that's when I see that error show up in the JS console of the app I'm working on that is talking to Sandstorm (via an HTTP API with web key-based auth).

**Comment by paulproteus** [comment]
@elimisteve FWIW unless this is affecting you in a serious way, we'll probably leave it open for a little while as we prioritize other bugfixes, and get to it when we can.

Inversely, if it _is_ affecting you in a serious way, you should tell us so we can prioritize it accordingly. :)

**Comment by elimisteve** [comment]
@paulproteus I agree that this isn't a serious problem.

Future visitors: just do another API call right after one fails and gives you the above message; the next request should work.

**Comment by paulproteus** [comment]
BTW @elimisteve it turns out that this _is_ a serious problem for us :) because it turns out that we're relying on Sanstorm HTTP APIs for some of our own stuff. So good news for you is we'll probably try to fix it soon!
